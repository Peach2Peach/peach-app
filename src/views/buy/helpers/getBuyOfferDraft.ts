type GetBuyOfferDraftParams = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
}
export const getBuyOfferDraft = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
}: GetBuyOfferDraftParams): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
})
