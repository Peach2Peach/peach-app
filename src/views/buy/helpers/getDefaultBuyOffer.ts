type DefaultOfferProps = {
  minBuyAmount: number
  maxBuyAmount: number
  meansOfPayment?: MeansOfPayment
  kyc?: boolean
}
export const getDefaultBuyOffer = ({
  minBuyAmount,
  maxBuyAmount,
  meansOfPayment = {},
  kyc = false,
}: DefaultOfferProps): BuyOfferDraft => ({
  type: 'bid',
  creationDate: new Date(),
  lastModified: new Date(),
  amount: [minBuyAmount, maxBuyAmount],
  meansOfPayment,
  paymentData: {},
  releaseAddress: '',
  originalPaymentData: [],
  kyc,
})
