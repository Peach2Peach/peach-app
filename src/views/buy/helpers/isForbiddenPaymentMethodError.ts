import { isPaymentMethod } from '../../../utils/validation/isPaymentMethod'

export const isForbiddenPaymentMethodError = (
  errorMessage: string | null,
  errorDetails: unknown,
): errorDetails is PaymentMethod[] =>
  errorMessage === 'FORBIDDEN' && Array.isArray(errorDetails) && errorDetails.every(isPaymentMethod)
