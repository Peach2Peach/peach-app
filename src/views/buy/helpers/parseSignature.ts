const signaturePattern = /[a-zA-Z0-9/=+]{88}/u

export const parseSignature = (signature: string) => signature.match(signaturePattern)?.pop() || signature
