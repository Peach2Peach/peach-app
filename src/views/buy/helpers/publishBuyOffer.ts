import { publishPGPPublicKey } from '../../../init/publishPGPPublicKey'
import { saveOffer } from '../../../utils/offer'
import { postBuyOffer } from '../../../utils/peachAPI'

export const publishBuyOffer = async (offerDraft: BuyOfferDraft) => {
  let [result, err] = await postBuyOffer(offerDraft)

  if (err?.error === 'PGP_MISSING') {
    await publishPGPPublicKey()
    const response = await postBuyOffer(offerDraft)
    result = response[0]
    err = response[1]
  }
  if (result) {
    saveOffer({ ...offerDraft, ...result })
    return { offerId: result.id, isOfferPublished: true, errorMessage: null }
  }
  return { isOfferPublished: false, errorMessage: err?.error || 'POST_OFFER_ERROR' }
}
