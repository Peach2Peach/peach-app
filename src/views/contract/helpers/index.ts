export { isPSBTForBatch } from "./isPSBTForBatch";
export { shouldShowTradeStatusInfo } from "./shouldShowTradeStatusInfo";
export { verifyReleasePSBT as verifyPSBT } from "./verifyReleasePSBT";
