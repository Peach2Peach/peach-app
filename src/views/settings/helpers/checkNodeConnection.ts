import { Blockchain } from 'bdk-rn'
import { BlockChainNames, BlockchainElectrumConfig, BlockchainEsploraConfig } from 'bdk-rn/lib/lib/enums'
import { info } from '../../../utils/log'
import { parseError } from '../../../utils/result'
import { addProtocol } from '../../../utils/web'

const checkElectrumConnection = async (address: string, ssl: boolean) => {
  const config: BlockchainElectrumConfig = {
    url: addProtocol(address, ssl ? 'ssl' : 'tcp'),
    sock5: null,
    retry: 1,
    timeout: 5,
    stopGap: 1,
    validateDomain: false,
  }

  try {
    info('Checking electrum connection...')
    const blockchain = await new Blockchain().create(config, BlockChainNames.Electrum)
    await blockchain.getBlockHash()
    return { result: BlockChainNames.Electrum }
  } catch (e) {
    info('electrum connection failed')
    return { error: parseError(e) }
  }
}
const checkEsploraConnection = async (address: string, ssl: boolean) => {
  const config: BlockchainEsploraConfig = {
    baseUrl: addProtocol(address, ssl ? 'https' : 'http'),
    proxy: null,
    concurrency: 1,
    timeout: 5,
    stopGap: 1,
  }

  try {
    info('Checking esplora connection...')
    const blockchain = await new Blockchain().create(config, BlockChainNames.Esplora)
    await blockchain.getBlockHash()
    return { result: BlockChainNames.Esplora }
  } catch (e) {
    info('esplora connection failed')
    return { error: parseError(e) }
  }
}
const connectionChecks = [checkElectrumConnection, checkEsploraConnection]
export const checkNodeConnection = async (address: string, ssl = false) => {
  const errors: unknown[] = []
  for (const connectionCheck of connectionChecks) {
    // eslint-disable-next-line no-await-in-loop
    const { result, error } = await connectionCheck(address, ssl)
    if (result) return { result }
    errors.push(error)
  }
  return { error: errors.join('\n\n') }
}
