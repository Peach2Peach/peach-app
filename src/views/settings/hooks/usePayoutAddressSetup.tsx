import { useMemo } from 'react'

import { shallow } from 'zustand/shallow'
import { HelpIcon } from '../../../components/icons'
import { useHeaderSetup, useNavigation, useRoute, useValidatedState } from '../../../hooks'
import { useShowHelp } from '../../../hooks/useShowHelp'
import { useSettingsStore } from '../../../store/settingsStore'
import i18n from '../../../utils/i18n'

const addressRules = { bitcoinAddress: true, blockTaprootAddress: true, required: true }
const labelRules = { required: true }
const title = {
  refund: 'settings.refundAddress',
  payout: 'settings.payoutAddress',
}
export const usePayoutAddressSetup = () => {
  const route = useRoute<'payoutAddress'>()
  const { type } = route.params || {}
  const navigation = useNavigation()
  const [payoutAddress, setPayoutAddress, payoutAddressLabel, setPayoutAddressLabel] = useSettingsStore(
    (state) => [state.payoutAddress, state.setPayoutAddress, state.payoutAddressLabel, state.setPayoutAddressLabel],
    shallow
  )
  const [address, setAddress, addressValid, addressErrors] = useValidatedState(payoutAddress || '', addressRules)
  const [addressLabel, setAddressLabel, addressLabelValid, addressLabelErrors] = useValidatedState(
    payoutAddressLabel || '',
    labelRules
  )
  const isUpdated = address === payoutAddress && addressLabel === payoutAddressLabel
  const showHelp = useShowHelp('payoutAddress')

  useHeaderSetup(
    useMemo(
      () => ({
        title: i18n(title[type || 'payout']),
        icons: [{ iconComponent: <HelpIcon />, onPress: showHelp }],
      }),
      [showHelp, type]
    )
  )

  const save = () => {
    if (addressValid && addressLabelValid) {
      const addressChanged = payoutAddress !== address
      setPayoutAddress(address)
      setPayoutAddressLabel(addressLabel)

      if (type === 'payout' && addressChanged) navigation.replace('signMessage')
      if (type === 'refund' && addressChanged) navigation.goBack()
    }
  }

  return {
    type,
    address,
    setAddress,
    addressErrors,
    addressValid,
    addressLabel,
    setAddressLabel,
    addressLabelErrors,
    addressLabelValid,
    isUpdated,
    save,
  }
}
