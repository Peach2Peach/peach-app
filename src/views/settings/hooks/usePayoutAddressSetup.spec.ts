import { renderHook } from '@testing-library/react-native'
import { headerState, NavigationWrapper } from '../../../../tests/unit/helpers/NavigationWrapper'
import { usePayoutAddressSetup } from './usePayoutAddressSetup'

const useRouteMock = jest.fn(() => ({
  params: {},
}))
jest.mock('../../../hooks/useRoute', () => ({
  useRoute: () => useRouteMock(),
}))

describe('usePayoutAddressSetup', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should set up header correctly', () => {
    renderHook(usePayoutAddressSetup, { wrapper: NavigationWrapper })
    expect(headerState.header()).toMatchSnapshot()
  })
})
