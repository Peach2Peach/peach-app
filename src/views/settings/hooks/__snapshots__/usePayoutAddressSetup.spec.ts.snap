// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePayoutAddressSetup should set up header correctly 1`] = `
<Header
  icons={
    [
      {
        "color": "#099DE2",
        "id": "helpCircle",
        "onPress": [Function],
      },
    ]
  }
  title="custom payout wallet"
/>
`;
